test/qa/test_routing.py

```diff
@@ -27,86 +29,152 @@ def setup_function(function):
 def teardown_function(function):
     logging.log(data=info.collect())
     logging.log()
-    lib.set_routing("on")
 
 
-@pytest.mark.flaky(reruns=2, reruns_delay=90)
+TABLE = 205
+SUBNET_1 = "1.1.1.1"
+SUBNET_2 = "2.2.2.2"
+SUBNET_3 = "3.3.3.3"
+
+MSG_ROUTING_OFF =  "Routing is set to 'disabled' successfully."
+MSG_ROUTING_ON = "Routing is set to 'enabled' successfully."
+MSG_ROUTING_OFF_ALREADY =  "Routing is already set to 'disabled'."
+MSG_ROUTING_ON_ALREADY = "Routing is already set to 'enabled'."
+MSG_ROUTING_USED_BY_MESH = "Routing is currently used by Meshnet. Disable it first."
+
+
+@pytest.mark.parametrize("tech,proto,obfuscated", lib.TECHNOLOGIES)
+#@pytest.mark.flaky(reruns=2, reruns_delay=90)
 @timeout_decorator.timeout(40)
-def test_routing_on():
-    subnet1 = "1.1.1.1"
-    subnet2 = "2.2.2.2"
-    subnet3 = "3.3.3.3"
-    lib.add_subnet_to_allowlist(f"{subnet1}/32")
-    lib.add_subnet_to_allowlist(f"{subnet2}/32")
-    lib.add_subnet_to_allowlist(f"{subnet3}/32")
-    table = 205
-
-    with lib.ErrorDefer(sh.nordvpn.allowlist.remove.all):
-        with lib.ErrorDefer(sh.nordvpn.disconnect):
-            output = sh.nordvpn.connect()
-            assert lib.is_connect_successful(output)
-
-            rules = sh.ip.rule.show.table(table)
-            assert "fwmark" in rules
-            policyRoutes = sh.ip.route.show.table(table)
-            assert subnet1 in policyRoutes
-            assert subnet2 in policyRoutes
-            assert subnet3 in policyRoutes
-            assert "nordlynx" in policyRoutes
+def test_routing_enabled_connect(tech, proto, obfuscated):
+    lib.set_technology_and_protocol(tech, proto, obfuscated)
+
+    lib.add_subnet_to_allowlist(f"{SUBNET_1}/32")
+    lib.add_subnet_to_allowlist(f"{SUBNET_2}/32")
+    lib.add_subnet_to_allowlist(f"{SUBNET_3}/32")
+
+    with lib.Defer(sh.nordvpn.allowlist.remove.all):
+        with lib.Defer(sh.nordvpn.disconnect):
```

Comment:
Could we do multiple actions on a single defer instead of nesting them?

Reply:
I believe in that case we'd need to alter the lib.Defer itself, but it is possible